src/compiler/scala/tools/nsc/util/CharArrayReader.scala

```diff
@@ -15,6 +16,12 @@ trait CharArrayReaderData {
   /** The offset one past the last read character */
   var charOffset: Int = 0
 
+  /** the offset of the character following the token preceding this one */
+  final def curOffset: Int = this match {
+    case td: Scanners#TokenData => td.lastOffset
+    case _ => charOffset - 1
+  }
```

Comment:
This is pretty gross, but I couldn't get my head around the tangle of inheritance here. The second case will kick in when we're in a `CharArrayLookaheadReader`.

Reply:
Let's add as a comment in the code.

Reply:
Polluting the reader because of lame xml parser doesn't seem right. I tried to get a handle on it. Basically, regular scanner knows when a token ends, xml parser does not. So when reader skips a CR to land on NL, `curpos - 1` is not the last char of the xml element. Oh, it looks like nextToken is smart enough to back up over the CR.  nextToken is invoked on scanner.resume when coming back from xml.

I'll try to make that more regular for xml, if I can figure it out before XML support is dropped entirely.